mtp_api/apps/mtp_auth/models.py

```diff
@@ -19,3 +22,43 @@ class ApplicationUserMapping(TimeStampedModel):
 
     def __str__(self):
         return self.user.username
+
+
+class FailedLoginAttempt(TimeStampedModel):
+    user = models.ForeignKey('auth.User')
+    application = models.ForeignKey('oauth2_provider.Application')
+
+    class Meta:
+        ordering = ('-created',)
+
+    @classmethod
+    def is_locked_out(cls, user, client):
+        failed_attempts = FailedLoginAttempt.objects.filter(
+            user=user,
+            application=client,
+        )
+        failed_attempt_count = failed_attempts.count()
+        if not failed_attempt_count:
+            return False
+
+        last_failed_attempt = failed_attempts.first()
+        lockout_period = datetime.timedelta(seconds=settings.MTP_AUTH_LOCKOUT_LOCKOUT_PERIOD)
+        return failed_attempt_count >= settings.MTP_AUTH_LOCKOUT_COUNT and \
+            last_failed_attempt.created > now() - lockout_period
+
+    @classmethod
+    def delete_failed_attempts(cls, user, client):
+        cls.objects.filter(
+            user=user,
+            application=client,
+        ).delete()
+
+    @classmethod
+    def add_failed_attempt(cls, user, client):
+        FailedLoginAttempt.objects.create(
+            user=user,
+            application=client,
+        )
```

Comment:
yeah i thought that but wasn't sure a whole new class was really necessary

On 10 November 2015 at 11:09, Ian Brechin <notifications@github.com> wrote:

> In mtp_api/apps/mtp_auth/models.py
> <https://github.com/ministryofjustice/money-to-prisoners-api/pull/64#discussion_r44394414>
> :
>
> > +        return failed_attempt_count >= settings.MTP_AUTH_LOCKOUT_COUNT and \
> > +            last_failed_attempt.created > now() - lockout_period
> > +
> > +    @classmethod
> > +    def delete_failed_attempts(cls, user, client):
> > +        cls.objects.filter(
> > +            user=user,
> > +            application=client,
> > +        ).delete()
> > +
> > +    @classmethod
> > +    def add_failed_attempt(cls, user, client):
> > +        FailedLoginAttempt.objects.create(
> > +            user=user,
> > +            application=client,
> > +        )
>
> I think it would be better to have these as methods on the Manager object
> rather than as class methods on the Model. Just kind of tidier.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ministryofjustice/money-to-prisoners-api/pull/64/files#r44394414>
> .
>